sim/core/focus.go

```diff
@@ -48,23 +124,57 @@ func (fb *focusBar) CurrentFocus() float64 {
 	return fb.currentFocus
 }
 
-func (fb *focusBar) AddFocus(sim *Simulation, amount float64, metrics *ResourceMetrics) {
+func (fb *focusBar) NextFocusTickAt() time.Duration {
+	return fb.nextFocusTick
+}
+
+// Returns the rate of focus regen per second from melee haste.
+// Todo: Verify that this is actually how it works. Check simc code below
+// player_t::focus_regen_per_second =========================================
+// double player_t::focus_regen_per_second() const
+//
+//	{
+//	  double r = base_focus_regen_per_second * ( 1.0 / composite_attack_haste() );
+//	  return r;
+//	}
+func (fb *focusBar) FocusRegenPerTick() float64 {
+	ticksPerSecond := float64(time.Second) / float64(FocusTickDuration)
+	hastePercent := fb.unit.RangedSwingSpeed()
+	tick := fb.baseFocusPerSecond * hastePercent / ticksPerSecond
+	return tick
+}
+
+func (fb *focusBar) onFocusGain(sim *Simulation, crossedThreshold bool) {
+	if sim.CurrentTime < 0 {
+		return
+	}
```

Comment:
Does this make sense? If you have this check here any prepull actions that cost focus will leave you with no focus regen until the combat starts